providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/EventStreamObserver.java

```diff
@@ -58,7 +59,9 @@ public void onError(Throwable t) {
             this.cache.clear();
         }
         this.stateConsumer.accept(ProviderState.ERROR);
-        this.reconnectEventStream.run();
+        synchronized (this.sync) {
```

Comment:
I think we should also add this to the `onCompleted()` method

Reply:
@thisthat good point. `onComplete` is a last call similar to `onError`. So we must reinitialize the listener

Reply:
@liran2000 we do not have multi-thread model here. Let me try to explain, 

We start a thread to handle the stream observer [^1], this is the only extra thread that we create. This thread handles a retry mechanism [^2] and executes an asynchronous `EventStreamObserver`.

`EventStreamObserver` is an implementation of grpc `StreamObserver`. `StreamObserver` implementation is such that `onError` & `onCompleted` calls are called once, and `no further calls to any method are allowed` [^3]

We `notify` waiting for caller when dealing with `onError` & `onCompleted`. Given these are the last calls, only the freshly initialized  `EventStreamObserver` can call the `notify` if that too face with and error. 

[^1]: https://github.com/Kavindu-Dodan/java-sdk-contrib/blob/feat/refactor-flagd/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java#L83

[^2]: https://github.com/Kavindu-Dodan/java-sdk-contrib/blob/feat/refactor-flagd/providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/grpc/GrpcConnector.java#L119

[^3]: https://grpc.github.io/grpc-java/javadoc/io/grpc/stub/StreamObserver.html#onError(java.lang.Throwable)